cargo-near/src/abi.rs

```diff
@@ -71,6 +72,40 @@ pub(crate) fn generate_abi(
         anyhow::bail!("`near-sdk` dependency must have the `abi` feature enabled")
     }
 
+    let near_sdk_node = crate_resolve
+        .nodes
+        .iter()
+        .find(|node| node.id == near_sdk_dep.id)
+        .ok_or_else(|| {
+            anyhow::anyhow!(
+                "unable to locate the `near-sdk` package in the dependency graph, please report this issue at https://github.com/near/cargo-near"
+            )
+        })?;
+    let near_abi_dep = near_sdk_node
+        .deps
+        .iter()
+        .find(|dep| dep.name == "near_abi")
+        .and_then(|near_abi| {
+            crate_metadata
+                .raw_metadata
+                .packages
+                .iter()
+                .find(|pkg| pkg.id == near_abi.pkg)
+        })
+        .ok_or_else(|| anyhow::anyhow!("expected `near-sdk` to provide `near-abi`, please report this issue at https://github.com/near/cargo-near"))?;
+
+    let expected_abi_ver = &Version::parse(env!("CARGO_NEAR_ABI_VERSION"))?;
+    let actual_abi_ver = &near_abi_dep.version;
+    if expected_abi_ver.major != actual_abi_ver.major
+        || expected_abi_ver.minor != actual_abi_ver.minor
```

Comment:
Hm, should we be testing with the crate version? or the `schema_version`?

Reply:
The reason why this check is needed in the first place is because `ChunkedAbiEntry` returned from the dylib artifact has to conform to the `ChunkedAbiEntry` used by `cargo-near`, which depends on how Rust ABI views them. For example I highly doubt that Rust would allow you to interpret:

```rust
pub struct AbiMetadata {
    pub name: Option<String>,
    pub other: HashMap<String, String>,
}
```
and
```rust
pub struct AbiMetadata {
    pub name: Option<String>,
    pub version: Option<String>,
    pub other: HashMap<String, String>,
}
```
as the same object even if they both can belong to the same `schema_version` (extending metadata object does not bump the schema version obviously).

If we were to switch back to JSON-based ABI chunk generation, then we would be able only to check the `schema_version`.

Reply:
Yeah I understand that, suggesting we could consider returning `serde_json::Value` from the chunks, instead of the strict type. Then we can leave it up to the deserialization implementation to do the validation. Or what do you think?

But as-is works too.

Reply:
Hmm, that could work. In general, do you know what invariants you can depend on when working with low-level Rust ABI? For example, is it reasonable to assume that `serde_json::Value` is going to have the same ABI for all `major.minor` releases? Also, would I have to re-compile cargo-near whenever I install a new version of rustc?

Reply:
Not sure, honestly, best path to avoid inconsistencies later down the line would probably be to return JSON strings. And manually handle deserialization wherein we can validate the `schema_version`.

Reply:
Yeah, that's what I am thinking too right now. Any ideas on how to get `schema_version` from a crate though? Can we add like a custom cargo metadata key or something?

Reply:
So, `ChunkedAbiEntry` already includes `schema_version`.

Which means we can only do the compatibility check after the `Generating ABI` step.

Reply:
I especially like that `ChunkedAbiEntry` already does this compatibility check on deserialization.

https://github.com/near/near-abi-rs/blob/368245073aa5bf919ee30f9e6f2246f68eee1b40/near-abi/src/private.rs#L11-L12

So we get this for free.

Reply:
True, I will prepare an alternative PR then